### ISSUE TITLE
switch { case null: ... } fails to match `undefined` from JS interop

### ISSUE THREAD

**Issue by mdebbar** [dart-lang/sdk#51527]
I'm facing an issue where a JS `undefined` makes its way into Dart code and Dart doesn't handle it properly:

https://dartpad.dev/66017295266de29cb7bf82d0136c7bb7

Basically, an `undefined` value sneaks through static js interop into the Dart world, and then `switch { case null: ... }` doesn't catch it.

**Comment by rakudrama** [comment]
dart2js and DDC both 'skip' a JavaScript `undefined`.

Probably the simplest fix is to compile Dart `case null: S;` to JavaScript
```js
case null:
case void 0:
  S;
```

With all the work going on with advanced switch-statements with patterns it is worth making a legacy and 'modern' test to ensure whatever fix we choose stays working.

**Comment by rakudrama** [comment]
Test available: https://dart-review.googlesource.com/c/sdk/+/290350

**Comment by skreborn** [comment]
### Workarounds, and the reasons they work

Interestingly, checking against `null` in a simple `if` works just fine.

```dart
// produces the correct output when `color` is `undefined`
if (color == null) {
  print('But it was, in fact, null.');
}
```

This compiles to the following, which is pretty straightforward. In JavaScript, `==` denotes loose equality, in which `null` and `undefined` are indeed equal. Using strict equality (`===`) instead in the generated code would make this fail all the same.

```javascript
if (color == null) {
  core.print("But it was, in fact, null.");
}
```

Adding a seemingly useless guard to the `switch` `case` also resolves the issue, as it is now a pattern to match instead of a value.

```dart
switch (color) {
  // Default color is blue.
  case null when true: // <-- `undefined` ends up matching this one
  case 'blue':
    // Expected:
    print('The color was blue.');
    break;
  case 'green':
    print('The color was green.');
    break;
  default:
    // Actual:
    throw Exception('Unknown color: $color (runtimeType: ${color.runtimeType})');
}
```

The reason being that this is compiled to entirely different JavaScript code, no longer utilizing a native `switch` statement, but instead a series of `if` statements. The generated code has been simplified for brevity, but its meaning is unchanged.

```javascript
if (color == null && true || color === "blue") {
    core.print("The color was blue.");
}
```

This is in opposition to the JavaScript code generated for the original example, which does, in fact, use a native `switch` statement.

```javascript
switch (color) {
  case null:
  case "blue":
      core.print("The color was blue.");
      break;
  // more cases...
}
```

All of these are applicable to both `dartdevc` and and `dart2js`.

### Differentiating between `null` and `undefined`

When I need to tell `null` and `undefined` values apart - which admittedly isn't often, but it _has_ happened -, I reach for this handy extension on `Object?`.

```dart
extension on Object? {
  bool get isUndefined => js_util.typeofEquals<Object?>(this, 'undefined');
}
```

Using it is as straightforward as you might imagine.

```dart
if (color == null) {
  if (color.isUndefined) {
    print("It is actually undefined!");
  } else {
    print("It is but a mere null.");
  }
}
```

This compiles to the following JavaScript code, which, again, is as straightforward as it gets (if you're familiar with the `typeof` expression). Once more, I've simplified the actual code for brevity, but I haven't changed its meaning.

```javascript
if (color == null) {
  if (typeof color == "undefined") {
    core.print("It is actually undefined!");
  } else {
    core.print("It is but a mere null.");
  }
}
```

**Comment by rakudrama** [comment]
We have fixed the issue (cc5b77a71045302005743d5ae871974dcc679423, 5a595124b47efadbebf95146a20d4645864fd93b)

@skreborn Thank you for your nicely written and detailed work-arounds.
